File: 39-combination-sum/39-combination-sum.java
class Solution {
    protected void backtrack(List<List<Integer>> results, LinkedList<Integer> comb, int[] candidates, int remain, int start) {
        if (remain == 0) {
            results.add(new ArrayList<Integer>(comb));
        }
        else if (remain < 0) {
            return;
        }
        for (int i = start; i < candidates.length; i++) {
            comb.add(candidates[i]);
            this.backtrack(results, comb, candidates, remain - candidates[i], i);
            comb.removeLast();
        }
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> results = new ArrayList();
        LinkedList<Integer> comb = new LinkedList();
        this.backtrack(results, comb, candidates, target, 0);
        return results;
    }
}

File: 39-combination-sum/39-combination-sum.py
class Solution:
        
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        combination = []
        sol = []
        amount = 0
        def backtrack(index, amount, combination):
            if index >= len(candidates) or (amount > target):
                return
            if (amount == target):
                sol.append(combination.copy())
                return
            #decision tree: include current element or skip it and don't add anything
            combination.append(candidates[index])
            amount += candidates[index]
            backtrack (index, amount, combination)
            combination.pop()
            amount -= candidates[index]
            backtrack(index + 1, amount, combination)
        backtrack(0, 0, [])
        return sol